Set/get parameters input widget values.

YooSunYoung opened the pull request:
Related to #140 
There are new getter and setter for input widget values.


# Setting widget values

https://github.com/user-attachments/assets/1e79105c-79e7-4cf6-b9d1-30f24e5bb430


# Getting widget values

https://github.com/user-attachments/assets/a862b4f7-cedb-4931-aaa2-f35fa9aa1e26

YooSunYoung commented:
It should also handle optional/switchable state... I'll do it in another PR....

YooSunYoung commented on src/ess/reduce/ui.py:
```diff
@@ -232,3 +234,136 @@ def refresh_workflow_box(change) -> None:
+        The keys are the parameter names or the type(Key in Sciline graphs).
+        The values are dictionaries with the following keys
+
+        - 'fields':
+            A dictionary of field names and values to set.
+            i.e. {'dim': 'tof', 'unit': 'mm'}
+            The field values will be set if the field exists in the widget.
+            Otherwise, the field value will be ignored and a warning will be raised.
+        - 'value':
+            If the widget has a value setter, the value will be set directly.
+            Otherwise, it will be ignored and a warning will be raised.
```
> Do I understand correctly that it is always either-or?

I guess...? But I wasn't entirely sure. We can assume they are.

> or are there other requirements I am not aware of? If it is about storing the "optional" flags etc., maybe it should be

Yes, we'll have to store optional/switchable widget status as well. I'll go for the second option.

SimonHeybrock replied:
If the value is a dictionary, isn't that the same as a widget having multiple fields?

YooSunYoung replied:
Not necessarily, a widget can decide how to make the `value` from `fields`.

For example, like this:

```python

class ROIWidget(GridBox, ValueWidget):
    def __init__(self):
        super().__init__()

        self.fields = {
            'xmin': FloatText(description='xmin'),
            'xmax': FloatText(description='xmax'),
            'ymin': FloatText(description='ymin'),
            'ymax': FloatText(description='ymax'),
            'unit': Text(description='unit'),
        }
        self.children = [
            Label(value="Select ROI:"),
            self.fields['xmin'],
            self.fields['xmax'],
            self.fields['ymin'],
            self.fields['ymax'],
            self.fields['unit'],
        ]

    @property
    def value(self):
        return {
            'xmin': sc.scalar(self.fields['xmin'].value, unit=self.fields['unit']),
            'xmax': sc.scalar(self.fields['xmax'].value, unit=self.fields['unit']),
            'ymin': sc.scalar(self.fields['ymin'].value, unit=self.fields['unit']),
            'ymax': sc.scalar(self.fields['ymax'].value, unit=self.fields['unit']),
        }
```

YooSunYoung replied:
This one is an example I just made up btw

SimonHeybrock replied:
1. Can we focus on our widgets and not on hypothetical situations?
2. We should distinguish between what the widget returns to the workflow (the `value`, your dict above) and the values of the fields. They don't need to be the same, so I am not sure what you example is demonstrating. Configuring the workflow would use `widget.value`. Saving/restoring the fields could use something like `widget.get_fields` and `widget.set_fields`. Using `widget.value` to get the field values isn't an option anyway, for widgets like the bin-edge widget, so whether there is a widget where `widget.value` returns a `dict` should be a moot point.

SimonHeybrock replied:
Are you suggesting to write `sc.vector([1,2,3], unit='m')`  in a text input widget and use `eval` on that?

Probably not, let us talk about this in person tomorrow or so.

YooSunYoung replied:
> or are there other requirements I am not aware of? If it is about storing the "optional" flags etc., maybe it should be

I have an idea how to handle this in wrapping widgets.
Just leaving a note so that I don't forget...

YooSunYoung commented on src/ess/reduce/ui.py:
```diff
@@ -232,3 +234,136 @@ def refresh_workflow_box(change) -> None:
+    parameter_box = _get_parameter_box(widget)
+    # Walk through the existing input widgets and set the values
+    for node, widget in parameter_box._input_widgets.items():
+        new_internal_values = new_parameter_values.get(
+            node, new_parameter_values.get(parameter_box._input_registry[node].name, {})
+        )
```
It first tries to find the `new_parameter_values` with a node(`type`) and if it can't find it it uses the `name` which is expected to be string version of type.

SimonHeybrock replied:
My question is why you have chosen such a design.

YooSunYoung replied:
Because there are a few scenarios of making these `new_parameter_values` dictionaries.
1. Hard-coding as a dictionary:
  It'd be more convenient to just copy+paste the names from the parameter box in this case.
2. Use workflow inputs or default parameter dictionaries from modules.
  Users will also want to use the results of the workflow for setting parameters as well,
  which gives a dictionary with keys as `type`
3. Mix of both.

SimonHeybrock replied:
Both 1.) and 2.) sound like it needs to be a type (Sciline key)? So should the string option be removed?

Remember it is always easier to add something later than to remove it. Any if you add support now you need to write (a) documentation and (b) more tests.

YooSunYoung replied:
> Both 1.) and 2.) sound like it needs to be a type (Sciline key)? So should the string option be removed?

I think that ruins the usability of this interface if we only allow types.
Then you need to know which module the types are from when you want to quickly update values from this interface.
So option 1) definitely needs string option...?

I think I'm confused. 
I thought the target users are people who don't want to use python interfaces directly and are using GUI instead.

If you can load the type for each target and everything, you won't need the GUI so much...?

SimonHeybrock replied:
Conclusion from discussion: Remove the string option for now.

YooSunYoung commented on src/ess/reduce/ui.py:
```diff
@@ -232,3 +234,136 @@ def refresh_workflow_box(change) -> None:
+        new_internal_values = new_parameter_values.get(
+            node, new_parameter_values.get(parameter_box._input_registry[node].name, {})
+        )
+        # Parse the new values and corresponding fields in the existing widget
+        new_fields: dict[str, Any] = new_internal_values.get('fields', {})
+        new_fields_keys = set(new_fields.keys())
+        widget_fields = getattr(widget, "fields", {})
+        # Extract valid fields
+        valid_fields = new_fields_keys & set(widget_fields.keys())
+        # Warn for invalid fields
+        invalid_fields = new_fields_keys - valid_fields
+        for field_name in invalid_fields:
+            warning_msg = f"Cannot set field '{field_name}' for parameter '{node}'."
+            " The field does not exist in the widget. "
+            "The field value will be ignored."
+            warnings.warn(warning_msg, UserWarning, stacklevel=1)
+        # Set the valid fields
+        for field_name in valid_fields:
+            widget_fields[field_name].value = new_fields[field_name]
+
+        # Set the high-level value if needed and possible.
+        # It is set after the fields to prioritize the high-level value.
+        if (self_value := new_internal_values.get('value')) is not None:
+            if _has_widget_value_setter(widget):
+                widget.value = self_value
+            else:
+                warning_msg = f"Cannot set value for parameter '{node}'. "
+                "The widget does not have a value setter. The value will be ignored."
+                warnings.warn(warning_msg, UserWarning, stacklevel=1)
```
Then we require all widgets to have `set_value` method.
It might also make it easier to handle optional/switchable wrapper... hmm... I'll try that though.

YooSunYoung commented on src/ess/reduce/ui.py:
```diff
@@ -232,3 +234,136 @@ def refresh_workflow_box(change) -> None:
+        }
+        if fields_values:  # Skip setting empty fields
+            input_widget_values['fields'] = fields_values
+        if _has_widget_value_setter(widget) and _is_value_exportable(widget.value):
+            input_widget_values['value'] = widget.value
```
I expected them to be stored in a json file so limited types of values that can be stored here.

SimonHeybrock replied:
Have you considered letting the code storing into JSON deal with that? If you "handle" this here it will also "break" all other applications which do not store to JSON.

YooSunYoung replied:
> it will also "break" all other applications which do not store to JSON.

I'm not sure what you mean by this.
In the updated implementation, the type of the value is not handled anymore.

YooSunYoung commented on src/ess/reduce/widgets/_base.py:
```diff
@@ -0,0 +1,63 @@
+
+@runtime_checkable
+class WidgetWithFieldsProtocol(Protocol):
+    def set_fields(self, new_values: dict[str, Any]) -> None: ...
+
+    def get_fields(self) -> dict[str, Any]: ...
```
I assumed that `set_fields` and `get_fields` should always exist in pairs
but let me know if there's any use-case that can only have one of them but not the other.

YooSunYoung commented:
@SimonHeybrock 
I updated the implementation as we discussed this morning but from the notebook it still looks similar so I didn't update the screen record.

Also it's not handling `OptionalWidget` or `SwitchableWidget` and `DropdownWidget` doesn't work currently.
I added them as a todo list in the issue #140

SimonHeybrock commented on src/ess/reduce/ui.py:
```diff
@@ -232,3 +234,80 @@ def refresh_workflow_box(change) -> None:
+    # ``node`s that don't exist in the input widgets will be ignored.
+    for node, widget in parameter_box._input_widgets.items():
+        if node in new_parameter_values:
+            # We shouldn't use `get` here because ``None`` is a valid value.
```
Which ipywidget can have a value of `None`?

YooSunYoung replied:
Because of `OptionalWidget`

SimonHeybrock commented on src/ess/reduce/ui.py:
```diff
@@ -232,3 +234,80 @@ def refresh_workflow_box(change) -> None:
+
+    Returns
+    -------
+    dict[str, Any]:
```
```suggestion
    :
```

SimonHeybrock commented on src/ess/reduce/ui.py:
```diff
@@ -232,3 +234,80 @@ def refresh_workflow_box(change) -> None:
+            set_fields(widget, new_parameter_values[node])
+
+
+def get_parameter_widget_values(
```
Should we try to find "symmetric" naming? Would be easier to remember and figure out that the output of one is the input for the other. We had `update_parameter_widget` above.

SimonHeybrock commented on src/ess/reduce/ui.py:
```diff
@@ -232,3 +234,80 @@ def refresh_workflow_box(change) -> None:
+) -> dict[type, Any]:
+    """Return the current values of the input widgets in the target widget.
+
+    The result of this function can be used to set the values of the input widgets
+    using the `~set_input_widget_values` function.
```
Outdated link? Was the function renamed?

SimonHeybrock commented on src/ess/reduce/widgets/_base.py:
```diff
@@ -0,0 +1,63 @@
+        widget.set_fields(new_values)
+    elif _has_widget_value_setter(widget):
+        widget.value = new_values
+    else:
+        warnings.warn(
+            f"Cannot set value or fields for widget of type {type(widget)}."
+            " The new_value(s) will be ignored.",
+            UserWarning,
+            stacklevel=1,
+        )
```
When can this happen? Is it just the cases that are not implemented yet? Otherwise I imagine all widgets can have their values set?

YooSunYoung replied:
No, it's for when you update the dictionary and made a typo so any field-names are incorrect.
I did it myself so it was useful for me...?

YooSunYoung replied:
Ah sorry I misunderstood the question it's found in `DropdownMenu`

SimonHeybrock reviewed (commented):
👍
